### ISSUE TITLE
Lost writes on table growth

### ISSUE THREAD

**Issue by puzpuzpuz** [cornelk/hashmap#54]
The map loses writes and it looks a lot (I didn't verify that) that the problem is in missing happens-before edges between growth and concurrent writes.

Most like this issue is the same as #46, but I didn't check their reproducer.

Here is a simple reproducer:
```go
func TestMapParallelStores(t *testing.T) {
	const (
		numStorers = 8
		numEntries = 10000
	)
	m := &HashMap{}
	cdone := make(chan bool)
	// Write to the map in parallel.
	for i := 0; i < numStorers; i++ {
		go func(id int) {
			for j := 0; j < numEntries; j++ {
				m.Set(strconv.Itoa(id)+":"+strconv.Itoa(j), j)
			}
			cdone <- true
		}(i)
	}
	// Wait for the goroutines to finish.
	for i := 0; i < numStorers; i++ {
		<-cdone
	}
	// Verify map contents.
	l := m.Len()
	if l != numStorers*numEntries {
		t.Errorf("invalid map length: got %d, expected %d", l, numStorers*numEntries)
	}
	for i := 0; i < numStorers; i++ {
		for j := 0; j < numEntries; j++ {
			k := strconv.Itoa(i) + ":" + strconv.Itoa(j)
			v, ok := m.GetStringKey(k)
			if !ok {
				t.Errorf("value not found for %s", k)
			}
			if vi, ok := v.(int); ok && vi != j {
				t.Errorf("values do not match for %s: %v", k, v)
			}
		}
	}
}
```

Environment: Ubuntu 20.04, Go 1.16.7.

Expected result: the test passes.
Note: the test may pass with a certain probability. For instance, the test always succeeds if `numEntries` set to 1 since no rehashing happens in such scenario.

Actual result: the test fails.
```bash
$ go test -count=10 -run=TestMapParallelStores
--- FAIL: TestMapParallelStores (3.79s)
    hashmap_test.go:700: value not found for 5:6070
...
--- FAIL: TestMapParallelStores (3.46s)
    hashmap_test.go:700: value not found for 4:3905
    hashmap_test.go:700: value not found for 7:4212
FAIL
exit status 1
FAIL	github.com/cornelk/hashmap	36.795s
```

**Comment by tigrato** [comment]
@puzpuzpuz in reality you do not lose any writes. The issue is when the table resizing is happening concurrently with writes the index lookup points to the wrong list position.

You can look into my fork which fixes that issue

**Comment by hello-100** [comment]
您好，您邮件已收，尽快给您回复。

**Comment by dgkanatsios** [comment]
@tigrato thanks for implementing the fix, I assume this is the fork: https://github.com/tigrato/hashmap

Are folks using this library or is there something else? @puzpuzpuz @danielporto

**Comment by hello-100** [comment]
您好，您邮件已收，尽快给您回复。

**Comment by cornelk** [comment]
With the fix https://github.com/cornelk/hashmap/pull/61 merged I was not able to reproduce this bug anymore.
Please review and create a new ticket if you see any other issue, thanks!
